src/Retry.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Retry
+  ( increment
+  , minDelay
+  , maxDelay
+  , RetryParams(RetryParams)
+  , retryQueue
+  , retryWithDelay
+  ) where
+
+import Data.Functor (($>))
+import Control.Concurrent (forkIO, threadDelay)
+import Control.Concurrent.Chan (newChan, readChan, writeChan)
+import Control.Exception (try)
+import System.Random (getStdRandom, randomR)
+
+-- This is growing closer to Control.Retry from the "retry" package.
+-- Some of this should probably be replaced with that.
+data RetryParams = RetryParams
+  { increment :: Float, -- Multiplier for how much longer to wait on each consecutive failure.  Eg: 1.5
+    minDelay :: Float,
+    maxDelay :: Float
+  }
+
+-- How long to wait
+nextBackoff :: RetryParams -> Bool -> Float -> Float
+nextBackoff params True _ = minDelay params
+nextBackoff params False prevBackoff =
+  max (minDelay params) (min (maxDelay params) (prevBackoff * increment params))
+
+retryWithDelay :: RetryParams -> IO Bool -> IO ()
+retryWithDelay params action = process $ minDelay params
+
+  where
+    process :: Float -> IO ()
+    process prevBackoff = (== Right True) <$> try @IOError action >>= \case
+      True -> pure ()
+      False -> do
+        let backoff = nextBackoff params False prevBackoff
+        delay <- getStdRandom (randomR (0, 2 * backoff))
+        threadDelay $ round $ 100000 * delay
+        process backoff
```

Comment:
This was the original code:

```haskell
    process prevBackoff = do
       success <- (== Right True) <$> (try action :: IO (Either IOError Bool))
       let backoff = nextBackoff params success prevBackoff
        in if success
             then pure ()
             else do
               delay <- getStdRandom (randomR (0, 2 * backoff))
               threadDelay $ round $ 100000 * delay
               process backoff
```

First, just a quick note that in a `do` block you don't need to include `in` here after a `let`.

I noticed that `backoff` only needs to be defined in the `else` branch here, and it seems better to just put it there. Also, seemed weird to me to used `success` in the definition of `backoff` since `backoff` is only used when `success` is `False`.

 The rest of my refactor here was motivated by my general policy of avoiding `do`, and preference for using case statements rather that if statements.